### PR TITLE
feat(util-user-agent-node): populate typescript version in user agent when available

### PR DESCRIPTION
**Pull request by trivikr**
### Issue
Internal JS-6661

### Description
Populates typescript version in user agent when available in `node_modules/typescript/package.json`

### Testing

This change is able to detect TypeScript version is CJS and ESM applications, as they both use CJS dependency
Details in https://github.com/trivikr/test-aws-sdk-js-typescript-detection

However, it's not able to detect TypeScript version in ESM bundles which don't populate `__dirname`, i.e. esbuild, rollup and rolldown. Details in https://github.com/trivikr/test-aws-sdk-js-typescript-detection/pull/5

It's able to detect TypeScript version in CJS bundles of popular bundlers, and ESM bundles of bundlers which populate `__dirname`, like webpack and rspack.

### Checklist
- [x] If the PR is a feature, add integration tests (`*.integ.spec.ts`) or E2E tests.
  - [ ] It's not a feature.
  - [x] Integration tests are written in https://github.com/trivikr/test-aws-sdk-js-typescript-detection, as they work from inside node_modules
- [ ] My E2E tests are resilient to concurrent i/o.
  - [x] I didn't write any E2E tests.
- [x] I added access level annotations e.g. `@public`, `@internal` tags and enabled doc generation on the package. Remember that access level annotations go below the description, not above.
  - [ ] I didn't add any public functions.
- [ ] Streams - how do they work?? My WebStream readers/locks are properly lifecycled. Node.js stream backpressure is handled. Error handling.
  - [x] No streams here.

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by trivikr** [comment]
Verified that bundling with rollup is successful, and it's able to detect typescript version

```js
// input.js
import { createDefaultUserAgentProvider } from "@aws-sdk/util-user-agent-node";

const userAgentProvider = createDefaultUserAgentProvider({
  serviceId: "s3",
  clientVersion: "0.1.0",
});
userAgentProvider().then(console.log);
```

```console
$ npx rollup input.js --file bundle.js --format esm

$ node bundle.js                         
[
  [ 'aws-sdk-js', '0.1.0' ],
  [ 'ua', '2.1' ],
  [ 'os/linux', '5.10.248-248.988.amzn2int.x86_64' ],
  [ 'lang/js' ],
  [ 'md/nodejs', '24.11.1' ],
  [ 'md/tsc', '5.9.3' ],
  [ 'api/s3', '0.1.0' ]
]
```

File: packages-internal/util-user-agent-node/src/getTypeScriptUserAgentPair.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import type { UserAgentPair } from "@smithy/types";
+import { readFile } from "node:fs/promises";
+
+import { getTypeScriptPackageJsonPath } from "./getTypeScriptPackageJsonPath";
+
+let tscVersion: string | null;
```
Comment (kuhe) [inline packages-internal/util-user-agent-node/src/getTypeScriptUserAgentPair.ts]:
```suggestion
let tscVersion: string | null | undefined;
```

File: packages-internal/util-user-agent-node/src/getTypeScriptPackageJsonPath.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { join, normalize, sep } from "node:path";
+
+export const getTypeScriptPackageJsonPath = (dirname = ""): string => {
```
Comment (kuhe) [inline packages-internal/util-user-agent-node/src/getTypeScriptPackageJsonPath.ts:11]:
```suggestion
/**
 * @internal
 */
export const getTypeScriptPackageJsonPath = (dirname = ""): string => {
```

Comment (trivikr) [reply packages-internal/util-user-agent-node/src/getTypeScriptPackageJsonPath.ts:11]: Added in https://github.com/aws/aws-sdk-js-v3/pull/7786/changes/dfdf947d7aa5f28765eebe23a44e90b0ae5d5f80

File: packages-internal/util-user-agent-node/src/getTypeScriptUserAgentPair.ts (hunk 1 above)
Comment (trivikr) [reply packages-internal/util-user-agent-node/src/getTypeScriptUserAgentPair.ts]: Not needed as tscVersion is either assigned a null or string explicitly.

Comment (kuhe) [reply packages-internal/util-user-agent-node/src/getTypeScriptUserAgentPair.ts]: it's initialized to undefined though, this could be 

```js
let tscVersion: string | null = "";
```

File: packages-internal/util-user-agent-node/src/getTypeScriptUserAgentPair.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+export const getTypeScriptUserAgentPair = async (): Promise<UserAgentPair | undefined> => {
+  // If tscVersion is set from previous calls.
+  if (tscVersion === null) {
+    return undefined;
+  } else if (tscVersion) {
+    return ["md/tsc", tscVersion];
+  }
+
+  try {
+    const packageJson = await readFile(getTypeScriptPackageJsonPath(__dirname), "utf-8");
```
Comment (kuhe) [inline packages-internal/util-user-agent-node/src/getTypeScriptUserAgentPair.ts:21]: extremely unlikely edge case: file is extremely large and causes readFile utf8 to run out of memory

I don't think you need to handle this case, since it's too unlikely

File: packages-internal/util-user-agent-node/src/getTypeScriptUserAgentPair.ts (hunk 1 above)
Comment (trivikr) [reply packages-internal/util-user-agent-node/src/getTypeScriptUserAgentPair.ts]: Added `undefined` type in https://github.com/aws/aws-sdk-js-v3/pull/7786/changes/9d670c16d6f4c8582828a4de03add2fed916fa23, since it's automatically set to undefined during initialization.

**Comment by trivikr** [comment]
The fix for the edge cases was tested in https://github.com/trivikr/test-aws-sdk-js-typescript-detection/pull/8
